### PR TITLE
adjust desugaring for async fn to correct drop order

### LINKED ISSUE #64391: async functions fail to move values into generated futures in the latest nightly
This simple async function fails to compile on the newest nightly: https://github.com/sfackler/rust-postgres/blob/2cc5bbf21b654c26bfd8b1e80b1e5c7cbd81235f/tokio-postgres/src/lib.rs#L163-L172

```
error[E0597]: `config` does not live long enough
   --> tokio-postgres/src/lib.rs:171:5
    |
171 |     config.connect(tls).await
    |     ^^^^^^-------------
    |     |
    |     borrowed value does not live long enough
    |     a temporary with access to the borrow is created here ...
172 | }
    | -
    | |
    | `config` dropped here while still borrowed
    | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `impl core::future::future::Future`
    |
    = note: The temporary is part of an expression at the end of a block. Consider forcing this temporary to be dropped sooner, before the block's local variables are dropped. For example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block.
```

I'm guessing this was caused by #64292 (cc @davidtwco)

### LINKED ISSUE #64512: async fn drop order for temporaries in return expression does not match sync
This asynchronous function:

```rust
async fn foo_async(x: D, y: D) {
    helper_async(&D()).await
    //                ^^^ temporary D
}

async fn helper_async(v: &D) { }
```

will drop first y, then x, then the temporary D. However an equivalent synchronous function ([playground]) would drop the temporary, then y, then x. The problem is our desugaring for async fn, which looks something like this:

[playground]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=99694c2815ff09142b1cc51c10b29111

```rust
// async fn foo_async($parameter_patterns) { $body }
fn foo_async(raw_parameters) {
  async move {
    let $parameter_patterns = raw-parameters;
    $body // body of the async fn
  }
}
```

here, the temporaries in `$body` wind up being dropped *after* the let-bound variables of the block. Proposed fix in a comment below.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Old desugaring, given a user function body `{ $stmts; $expr }`

```
{
    let $param_pattern0 = $raw_param0;
    ...
    let $param_patternN = $raw_paramN;
    $stmts;
    $expr
}
```

New desugaring:

```
{
    let $param_pattern0 = $raw_param0;
    ...
    let $param_patternN = $raw_paramN;
    drop-temps {
        $stmts;
        $expr
    }
}
```

The drop-temps is an internal bit of HIR that drops temporaries from the resulting expression, but it should be equivalent to `return { $stmts; $expr }`.

Fixes #64512 
Fixes #64391

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@cramertj is on vacation -- maybe @davidtwco feels up to reviewing this?

r? @davidtwco 

If you don't, though, let me know and we'll find somebody else. =)

**Comment by estebank** [comment]
This looks correct to me. If no reviewer gets a chance to see it in a timely manner, r=me.

File: src/librustc/hir/lowering/item.rs [hunk 1]
```diff
@@ -1219,8 +1219,44 @@ impl LoweringContext<'_> {
             let async_expr = this.make_async_expr(
                 CaptureBy::Value, closure_id, None, body.span,
                 |this| {
-                    let body = this.lower_block_with_stmts(body, false, statements);
-                    this.expr_block(body, ThinVec::new())
+                    // Create a block from the user's function body:
+                    let user_body = this.lower_block(body, false);
+                    let user_body = this.expr_block(user_body, ThinVec::new());
```
Comment (Centril) [inline src/librustc/hir/lowering/item.rs]: This `lower_block` + `expr_block` happens in the lowering of:

- `ExprKind::Async`
- `lower_expr_if`
- `lower_expr_while_in_loop_scope`

Introduce `fn lower_block_expr` and refactor those?

File: src/librustc/hir/lowering/item.rs [hunk 2]
```diff
@@ -1219,8 +1219,44 @@ impl LoweringContext<'_> {
...
+                        ThinVec::new(),
+                    );
+
+                    // Create a block like
+                    //
+                    // ```
+                    // {
+                    //   let $param_pattern = $raw_param;
+                    //   ...
+                    //   drop-temps { <user-body> }
```
Comment (Centril) [inline src/librustc/hir/lowering/item.rs]: Notation used in `fn lower_expr_while_in_loop_scope` is `match DropTemps($cond) {`. (Use one of the notations consistently... yours is fine ^^)

File: src/librustc/hir/lowering/item.rs [hunk 3]
```diff
@@ -1219,8 +1219,44 @@ impl LoweringContext<'_> {
...
+                        user_body.span,
+                        None,
+                    );
+                    let user_body = this.expr_drop_temps(
+                        desugared_span,
+                        P(user_body),
+                        ThinVec::new(),
+                    );
+
+                    // Create a block like
```
Comment (Centril) [inline src/librustc/hir/lowering/item.rs]: Seems like the larger description from/into needs to be updated as well?

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
+// aux-build:arc_wake.rs
+// edition:2018
+// run-pass
+
+#![allow(unused_variables)]
+
+// Test that the drop order for parameters in a fn and async fn matches up. Also test that
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
// Test that the drop order for parameters in an `fn` and `async fn` matches up. Also test that
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+// parameters (used or unused) are not dropped until the async fn completes execution.
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
// parameters (used or unused) are not dropped until the `async fn` completes execution.
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+struct D(&'static str, DropOrderListPtr);
+
+impl Drop for D {
+    fn drop(&mut self) {
+        self.1.borrow_mut().push(DropOrder::Val(self.0));
+    }
+}
+
+/// Check drop order of temporary "temp" as compared to x, y, and z.
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
/// Check drop order of temporary "temp" as compared to `x`, `y`, and `z`.
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+///
+/// Expected order:
+/// - z
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
/// - `z`
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+/// - temp
+/// - y
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
/// - `y`
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+/// - x
```
Comment (Centril) [inline src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]:
```suggestion
/// - `x`
```

File: src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs (hunk 4 above)
Comment (nikomatsakis) [reply src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]: "an" seems wrong here, fyi

Comment (nikomatsakis) [reply src/test/ui/async-await/drop-order/drop-order-for-temporary-in-tail-return-expr.rs]: I rewrote the description completely

**Review by davidtwco — APPROVED** [review]
LGTM! r=me when CI is passing.

**Review by davidtwco — COMMENTED** [review]
Oh, just remembered, will the reference need updated with this change? rust-lang-nursery/reference#674 is still open from the initial change.

**Comment by nikomatsakis** [comment]
@davidtwco 

> Oh, just remembered, will the reference need updated with this change? rust-lang-nursery/reference#674 is still open from the initial change.

I was planning to take a look at the reference and decide what needed to be updated.

**Pull request merged**